packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart

```diff
@@ -136,6 +143,69 @@ final class AssetTransformer {
   }
 }
 
+
+// A wrapper around [AssetTransformer] to support hot reload of transformed assets.
+final class DevelopmentAssetTransformer {
+  DevelopmentAssetTransformer({
+    required FileSystem fileSystem,
+    required AssetTransformer transformer,
+    required Logger logger,
+  })  : _fileSystem = fileSystem,
+        _transformer = transformer,
+        _logger = logger;
+
+  final AssetTransformer _transformer;
+  final FileSystem _fileSystem;
+  final Pool _transformationPool = Pool(4);
+  final Logger _logger;
+
+  /// Re-transforms an asset and returns a [DevFSContent] that should be synced
+  /// to the attached device in its place.
+  ///
+  /// Returns `null` if any of the transformation subprocesses failed.
+  Future<DevFSContent?> retransformAsset({
+    required String inputAssetKey,
+    required DevFSContent inputAssetContent,
+    required List<AssetTransformerEntry> transformerEntries,
+    required String workingDirectory,
+  }) async {
+    final File output = _fileSystem.systemTempDirectory.childFile('retransformerInput-$inputAssetKey');
+    ErrorHandlingFileSystem.deleteIfExists(output);
+    late File inputFile;
```

Comment:
`late` should generally be avoided (except it's useful in test code, where you ensure it's always initialized from `setUp()`). Consider if line 172 throws and we never initialize `inputFile`, we'd get a `LateInitializationError` on line 202. Maybe we make it nullable, and then update the check on 201 to be `if (inputFile != null && cleanupInput)`?

I think it's the right tool for the job here of unconditionally cleaning up resources, however `finally` can make it more difficult to reason about the type safety of a block of code.

Reply:
Addressed in [nit](https://github.com/flutter/flutter/pull/144161/commits/37bb741ba64a22b957d1ef1e7ded27c7e7ea11ad), but in slightly a different way than proposed. I make a null-assertion on `inputFile`, since it should be set if `cleanupInput` was set.